File: tests/vault_tests.rs
use anchor_lang::{prelude::Pubkey, system_program, AnchorDeserialize, InstructionData, ToAccountMetas};
use litesvm::LiteSVM;
use solana_sdk::{
    instruction::Instruction,
    signature::{Keypair, Signer},
    transaction::Transaction,
    program_pack::Pack,
};
use spl_token::state::{Account as TokenAccount, Mint};
use spl_associated_token_account::get_associated_token_address;


// helpers

fn program_id()-> Pubkey {
    spl_vault_anchor::ID
}

fn vault_state_pda(admin: &Pubkey) -> (Pubkey, u8){
    Pubkey::find_program_address(&[b"vault_state", admin.as_ref()], &program_id())
}

fn vault_token_pda(vault_state: &Pubkey) -> (Pubkey, u8){
    Pubkey::find_program_address(&[b"vault_token", vault_state.as_ref()], &program_id())
}

fn withdrawal_ticket_pda(user: &Pubkey, vault_state: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"withdrawal", user.as_ref(), vault_state.as_ref()],
        &program_id(),
    )
}

//create a mint, return its pubkey
fn create_mint(svm: &mut LiteSVM, payer: &Keypair, decimals: u8) -> Pubkey {
    let mint_kp = Keypair::new();
    let rent = svm.minimum_balance_for_rent_exemption(Mint::LEN);
    let create_ix = solana_sdk::system_instruction::create_account(
        &payer.pubkey(),
        &mint_kp.pubkey(),
        rent,
        Mint::LEN as u64,
        &spl_token::id(),
    );

    let init_ix = spl_token::instruction::initialize_mint(
        &spl_token::id(),
        &mint_kp.pubkey(),
        &payer.pubkey(),
        None,
        decimals,
    ).unwrap();

    let tx = Transaction::new_signed_with_payer(
        &[create_ix, init_ix],
        Some(&payer.pubkey()),
        &[payer, &mint_kp],
        svm.latest_blockhash(),
    );
    svm.send_transaction(tx).unwrap();
    mint_kp.pubkey()
}

//create an ata for owner, return its pubkey
fn create_ata(svm: &mut LiteSVM, payer: &Keypair, owner: &Pubkey, mint: &Pubkey) -> Pubkey {
    let ata = get_associated_token_address(owner, mint);
    let ix = spl_associated_token_account::instruction::create_associated_token_account(
        &payer.pubkey(),
        owner,
        mint,
        &spl_token::id(),
    );

    let tx = Transaction::new_signed_with_payer(
        &[ix],
        Some(&payer.pubkey()),
        &[payer],
        svm.latest_blockhash(),
    );

    svm.send_transaction(tx).unwrap();
    ata
}

//mint tokens to a token acc
fn mint_to(svm: &mut LiteSVM, payer: &Keypair, mint: &Pubkey, dest: &Pubkey, amount: u64) {
    let ix = spl_token::instruction::mint_to(
        &spl_token::id(),
        mint,
        dest,
        &payer.pubkey(),
        &[],
        amount,
    ).unwrap();

    let tx = Transaction::new_signed_with_payer(
        &[ix],
        Some(&payer.pubkey()),
        &[payer],
        svm.latest_blockhash(),
    );

    svm.send_transaction(tx).unwrap();
}

//read token acc bal
fn token_balance(svm: &LiteSVM, account: Pubkey) -> u64 {
    let data = svm.get_account(account).unwrap().data;
    TokenAccount::unpack(&data).unwrap().amount
}

// setup

struct TestContext {
    svm: LiteSVM,
    admin: Keypair,
    user: Keypair,
    accepted_mint: Pubkey,
    receipt_mint_kp: Keypair,
    vault_state: Pubkey,
    vault_token_account: Pubkey,
    user_token_ata: Pubkey,
    user_receipt_ata: Pubkey,
    admin_token_ata: Pubkey,
}
impl TestContext {
    fn new() -> Self {
        let mut svm = LiteSVM::new();
        svm.add_program_from_file(program_id(), "../../target/deploy/spl_vault_anchor.so").unwrap();

        let admin = Keypair::new();
        let user = Keypair::new();
        svm.airdrop(&admin.pubkey(), 10_000_000_000).unwrap();
        svm.airdrop(&user.pubkey(), 10_000_000_000).unwrap();

        // create accepted mint (admin is mint authority)
        let accepted_mint = create_mint(&mut svm, &admin, 6);

        //receipt mint keupair (passed to initialize)
        let receipt_mint_kp = Keypair::new();

        // derive pdas
        let (vault_state, _) = vault_state_pda(&admin.pubkey());
        let (vault_token_account, _) = vault_token_pda(&vault_state);

        //create atas
        let user_token_ata = create_ata(&mut svm, &admin, &user.pubkey(), &accepted_mint);
        let admin_token_ata = create_ata(&mut svm, &admin, &admin.pubkey(), &accepted_mint);

        //mint 10,000 tokrn to user
        mint_to(&mut svm, &admin, &accepted_mint, &user_token_ata, 10_000);

        TestContext {
            svm,
            admin,
            user,
            accepted_mint,
            receipt_mint_kp,
            vault_state,
            vault_token_account,
            user_token_ata,
            user_receipt_ata: Pubkey::default(), //set after initialize
            admin_token_ata,
        }
    }

    //call initialize and set user_receipt_ata
    fn initialize(&mut self, fee_bps: u16){
        let ix = Instruction {
            program_id: program_id(),
            accounts: spl_vault_anchor::accounts::Initialize {
                admin: self.admin.pubkey(),
                accepted_mint: self.accepted_mint,
                receipt_mint: self.receipt_mint_kp.pubkey(),
                vault_state: self.vault_state,
                vault_token_account: self.vault_token_account,
                token_program: spl_token::id(),
                system_program: system_program::ID,
                rent: solana_sdk::sysvar::rent::id(),
            }
            .to_account_metas(None),
            data: spl_vault_anchor::instruction::Initialize{fee_bps}.data(),
        };

        let tx = Transaction::new_signed_with_payer(
            &[ix],
            Some(&self.admin.pubkey()),
            &[&self.admin, &self.receipt_mint_kp],
            self.svm.latest_blockhash(),
        );
        self.svm.send_transaction(tx).unwrap();

        //create user receip ata
        self.user_receipt_ata = create_ata(
            &mut self.svm,
            &self.admin,
            &self.user.pubkey(),
            &self.receipt_mint_kp.pubkey(),
        );
    }
}

File: tests/vault_tests.ts
import * as anchor from "@coral-xyz/anchor";
import { Program, BN } from "@coral-xyz/anchor";
import { SplTokenVault } from "../target/types/spl_token_vault";
import {
  createMint,
  createAccount,
  mintTo,
  getAccount,
  TOKEN_PROGRAM_ID,
} from "@solana/spl-token";
import { assert } from "chai";

describe("spl-vault-anchor", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.SplTokenVault as Program<SplTokenVault>;
  const connection = provider.connection;

  // keypairs
  const admin = anchor.web3.Keypair.generate();
  const user = anchor.web3.Keypair.generate();
  let acceptedMint: anchor.web3.PublicKey;
  let receiptMint: anchor.web3.Keypair;
  let vaultState: anchor.web3.PublicKey;
  let vaultTokenAccount: anchor.web3.PublicKey;
  let userTokenAccount: anchor.web3.PublicKey;
  let userReceiptAccount: anchor.web3.PublicKey;
  let adminTokenAccount: anchor.web3.PublicKey;

  const FEE_BPS = 50; // 0.5%
  const DEPOSIT_AMOUNT = new BN(1_000);
  const COOLDOWN_SECONDS = 86_400;

  before(async () => {
    // airdrop to admin and user
    await connection.confirmTransaction(
      await connection.requestAirdrop(admin.publicKey, 10 * anchor.web3.LAMPORTS_PER_SOL)
    );
    await connection.confirmTransaction(
      await connection.requestAirdrop(user.publicKey, 10 * anchor.web3.LAMPORTS_PER_SOL)
    );

    // create accepted mint (admin is mint authority)
    acceptedMint = await createMint(connection, admin, admin.publicKey, null, 6);

    // receipt mint keypair (passed to initialize as signer)
    receiptMint = anchor.web3.Keypair.generate();

    // derive PDAs
    [vaultState] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("vault_state"), admin.publicKey.toBuffer()],
      program.programId
    );
    [vaultTokenAccount] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("vault_token"), vaultState.toBuffer()],
      program.programId
    );

    // create token accounts
    userTokenAccount = await createAccount(connection, user, acceptedMint, user.publicKey);
    adminTokenAccount = await createAccount(connection, admin, acceptedMint, admin.publicKey);

    // mint 10,000 tokens to user
    await mintTo(connection, admin, acceptedMint, userTokenAccount, admin, 10_000);
  });

  it("initializes the vault", async () => {
    await program.methods
      .initialize(FEE_BPS)
      .accounts({
        admin: admin.publicKey,
        acceptedMint,
        receiptMint: receiptMint.publicKey,
        vaultState,
        vaultTokenAccount,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .signers([admin, receiptMint])
      .rpc();

    const vault = await program.account.vaultState.fetch(vaultState);
    assert.ok(vault.admin.equals(admin.publicKey));
    assert.equal(vault.totalDeposited.toNumber(), 0);
    assert.equal(vault.feeBps, FEE_BPS);
    assert.equal(vault.isPaused, false);

    // create user receipt ATA now that receipt mint exists
    userReceiptAccount = await createAccount(
      connection,
      user,
      receiptMint.publicKey,
      user.publicKey
    );

    console.log("✅ Vault initialized correctly");
  });

  it("deposits tokens and mints receipts 1:1", async () => {
    await program.methods
      .deposit(DEPOSIT_AMOUNT)
      .accounts({
        user: user.publicKey,
        vaultState,
        userTokenAccount,
        vaultTokenAccount,
        receiptMint: receiptMint.publicKey,
        userReceiptAccount,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .signers([user])
      .rpc();

    const vaultAcc = await getAccount(connection, vaultTokenAccount);
    assert.equal(Number(vaultAcc.amount), 1_000);

    const receiptAcc = await getAccount(connection, userReceiptAccount);
    assert.equal(Number(receiptAcc.amount), 1_000);

    const vault = await program.account.vaultState.fetch(vaultState);
    assert.equal(vault.totalDeposited.toNumber(), 1_000);

    console.log("✅ Deposit minted receipts 1:1");
  });

  it("requests withdrawal — burns receipts and creates ticket", async () => {
    const [withdrawalTicket] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("withdrawal"), user.publicKey.toBuffer(), vaultState.toBuffer()],
      program.programId
    );

    await program.methods
      .requestWithdrawal(DEPOSIT_AMOUNT)
      .accounts({
        user: user.publicKey,
        vaultState,
        receiptMint: receiptMint.publicKey,
        userReceiptAccount,
        withdrawalTicket,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .signers([user])
      .rpc();

    // receipts should be burned
    const receiptAcc = await getAccount(connection, userReceiptAccount);
    assert.equal(Number(receiptAcc.amount), 0);

    // ticket should exist with correct data
    const ticket = await program.account.withdrawalTicket.fetch(withdrawalTicket);
    assert.equal(ticket.receiptAmount.toNumber(), 1_000);
    assert.ok(ticket.user.equals(user.publicKey));

    console.log("✅ Withdrawal ticket created, receipts burned");
  });

  it("rejects claim before 24hr cooldown", async () => {
    const [withdrawalTicket] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("withdrawal"), user.publicKey.toBuffer(), vaultState.toBuffer()],
      program.programId
    );

    try {
      await program.methods
        .claim()
        .accounts({
          user: user.publicKey,
          vaultState,
          vaultTokenAccount,
          userTokenAccount,
          adminTokenAccount,
          withdrawalTicket,
          tokenProgram: TOKEN_PROGRAM_ID,
        })
        .signers([user])
        .rpc();
      assert.fail("Should have thrown CooldownNotElapsed");
    } catch (err: any) {
      assert.include(err.message, "CooldownNotElapsed");
      console.log("✅ Claim correctly rejected before 24hr cooldown");
    }
  });

  it("attacker cannot claim another user's ticket", async () => {
    const attacker = anchor.web3.Keypair.generate();
    await connection.confirmTransaction(
      await connection.requestAirdrop(attacker.publicKey, 2 * anchor.web3.LAMPORTS_PER_SOL)
    );

    // real ticket belongs to user, attacker tries to claim it
    const [realTicket] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("withdrawal"), user.publicKey.toBuffer(), vaultState.toBuffer()],
      program.programId
    );

    try {
      await program.methods
        .claim()
        .accounts({
          user: attacker.publicKey,
          vaultState,
          vaultTokenAccount,
          userTokenAccount,
          adminTokenAccount,
          withdrawalTicket: realTicket,
          tokenProgram: TOKEN_PROGRAM_ID,
        })
        .signers([attacker])
        .rpc();
      assert.fail("Attacker should not be able to claim user's ticket");
    } catch (err: any) {
      // expected — PDA seeds won't match attacker's pubkey
      console.log("✅ Attacker correctly rejected from stealing withdrawal");
    }
  });
});
